### DISCUSSION TITLE
Plugin by Dev /// Dev home pages

### DISCUSSION THREAD

**Discussion by psobolewskiPhD** [chanzuckerberg/napari-hub#299] [Ideas]
From the main page, one can search by Dev, which is nice, if you know in advance that say Robert Haase @haesleinhuepf is a superstar plugin author.
But if you're on a plugin page, like: https://www.napari-hub.org/plugins/napari-crop
you can't easily get to Robert's other plugins, as far as I can tell. 
My expectation would be—at the least—that here:
<img width="634" alt="image" src="https://user-images.githubusercontent.com/76622105/138518898-25f05ecf-f2d2-471f-8d59-3e7f0991b098.png">

Under Authors: the `Robert Haase` would be a clickable link, like all the others to the right.

To go further, I think it would be a good idea to have Dev profiles/pages which would help validate plugins. I mean, as far as I can tell, anyone—even me—can make a `napari-crop` plugin. in fact, I have a Gist that uses *zoom* for this! Derp! Obviously I'm bad, but folks like @haesleinhuepf with **many plugins** need to be highlighted in some way! If someone has multiple plugins that's a major form of verification, because some troll isn't going to work so hard.

**Comment by psobolewskiPhD** [comment]
To add to this, based on a twitter thread with @haesleinhuepf (https://twitter.com/haesleinhuepf/status/1451650535027793923?s=21), I now realize it's possible to make `meta` plugins, that are really just collections of plugins, such as: devbio-napari
It could be useful for an authoritative napari figure, such as a key Dev or plugin 🧙 like @tlambert03 or @haesleinhuepf, etc. to have/make such a plugin that combines a few quality-of-life (QoL) plugins like @haesleinhuepf histograms, life profiles, cropping, plus a reader like aicsimageio, and maybe a few more? clesperanto? I dunno, a starter pack of plugins that can help a newbie grok napari (and python behind the scenes). Sure, you can do most/all of that *programmatically*, but not *everyone* can...

**Reply by psobolewskiPhD** [reply]
So, maybe some sort of Twitter blue check could work? To indicate verified non-troll non-noob plugins?

**Reply by haesleinhuepf** [reply]
Coming back to the original discussion :-D

I think meta-plugins or plugin-collections should be maintained by others, not by plugin developers, because they have a conflict of interest. Again, Peter @psobolewskiPhD , if you think of founding a plugin-collection e.g. for quality-of-life, I would totally support you and show you how this can be achieved.

**Comment by haesleinhuepf** [comment]
> Sure, you can do most/all of that programmatically, but not everyone can...

I recently had some discussions with @sofroniewn, @jni and others who remineded me of the first sentence [on the website](https://napari.org/): "napari is a fast, interactive, multi-dimensional image viewer __for Python__." Thus, I have the impression non-programmers are not an important target audience of napari (anymore?). Hence, people like you @psobolewskiPhD and me, who are closer to that audience who lacks coding skills, should build the bridge between the communities. I'm suggesting: @psobolewskiPhD why don't you take care of a quality-of-life plugin collection for napari just like I maintain [devbio-napari](https://github.com/haesleinhuepf/devbio-napari)?  I'm happy to help when it comes to technical difficulties. I presume you are also less biased than me when it comes to selecting plugins from other developers and thus, better suited for that job ;-)

**Reply by tlambert03** [reply]
> I have the impression non-programmers are not an important target audience of napari (anymore?). Hence, people like you @psobolewskiPhD and me, who are closer to that audience who lacks coding skills, should build the bridge between the communities.

i fully reject both the characterization that non-programmers are not an important target audience of napari (_definitely_ not "not anymore"), and that you are somehow closer to that audience that napari developers are. 

The deal here is and has always been that: before you can begin to offer a safe, robust, and non-frustrating interface to non-coders, you have to have a very clear and robust understanding of what that program even _is_, how it's organized, and where its pitfalls are.

Despite appearing rather often on twitter, I still see napari as a very promising experiment (not as something ready for mass consumption if only someone would build that bridge).
Everyone that I have talked to on the napari team very much wants to bridge this gap as much as you do, I think we just have different ideas about how best to go about that.  I'm not interested in rushing into it, and creating what is ostensibly a friendly interface for non-coders, but which depends on potentially fragile experimental features of napari (dock widgets, etc), that have the potential to frustrate non coders.  I would argue that that "bridge" needs to be as robust if not more than the base... and these plugins often tend to be rather fragile at the moment (that's partially an issue with the plugin engine itself... but we're working hard on that).

Juan, I, and other devs, spend most of our non-coding time working with end-users with no programming skills just like you do.  It's absolutely fair to have different opinions on what those users need and want, or how best to deliver it to them, but i definitely don't think you should assume you're on your own "build the bridge between the communities"

I would hope that the early adopters would want to work closely with napari to identify places in which napari could do better for non coders.  (You both have already done that in many ways and it's greatly appreciated)... but I hope you won't be tempted to think "doesn't look like napari is interested in it, i guess we have to do it ourselves).

**Reply by psobolewskiPhD** [reply]
@tlambert03  wrote:
>I would argue that that "bridge" needs to be as robust if not more than the base... and these plugins often tend to be rather fragile at the moment (that's partially an issue with the plugin engine itself... but we're working hard on that).
> 

I think this is a really important point, that sort of relates what I was getting at. Totally agree that for a GUI user, where plugins provide key functions, they need to be robust, because any errors issues will be more confusing and frustrating than for a coder who is familiar with tracebacks, GitHub issues, etc. 
What I was trying to get at is facilitating some way of having validation that certain plugins (or collections, whatever) are robust (within reason), meaning they work and are supported (by someone atleast, not napari devs). So a Plugin Dev with multiple plugins? To me that's a good sign, a gold star or blue checkmark, whatevs. I realize some may be experimental or broken, but still it's someone who seems to have some idea of what's going on. A playlist of plugins used by a Napari Dev, as mentioned by @sofroniewn ? That works, again, someone who has a clue uses these. To me, who frequently *doesn't* have a clue, that's a good sign! Gold star. 
Just something to make the cream rise to the top.

**Reply by tlambert03** [reply]
yeah, the general concept of getting the cream to rise to the top is useful to think about.  I find the star system on github to be pretty helpful, so it could ultimately be as simple as a hive-mind kind of thing (though of course, that takes years of interaction to really start to reflect anything meaningful).

I'm sure @neuromusic and the hub team are thinking hard about this... and I'd love to hear their thoughts and long term strategies.

**Reply by neuromusic** [reply]
> I'd love to hear their thoughts and long term strategies.

It is high priority for us to start building out some mechanisms here to give insight into quality in its many dimensions next year. (see #244)... A persistent theme in our UXR and convos w/ users is the struggle to find relevant, high quality image analysis tools & this motivates a lot of our work on the napari hub. Its important to get right, though, otherwise its either not useful or (worst case) actively harmful. For example, I agree with @psobolewskiPhD's point "If someone has multiple plugins that's a major form of verification, because some troll isn't going to work so hard." But at the same time, we also want to make sure that the napari ecosystem continues to be welcoming to newcomers so it continues to grow and thrive. If we did it wrong, @haesleinhuepf will end up the only plugin developer on here 🤣. This isn't to say we should avoid such surfacing such signal... but we may want to balance it with features that support and highlight newcomers, too. And your point @tlambert03 on how many of these mechanisms take time to cultivate enough signal is spot-on & critical as we think about any metrics that we try to cultivate & crowd-source.

In addition to Github, two sites that I think do this well for their respective communities (but take VERY different approaches) are [Kaggle Datasets](https://www.kaggle.com/datasets) and [G2](https://www.g2.com/)

This is all to say, we're excited but being cautious here. This is why, when we launched, we made an explicit decision to limit what kinds of signal & metrics can be used for searching & filtering. For example, even though we can get Github stars easily enough (you can see them on individual pages), we don't support filtering or sorting by Github stars. We want to take time to talk with users and think through the tradeoffs and priorities between this and other signals into quality.

@LCObus is actively leading our UXR work here, with a goal of having some more concrete ideas & priorities by December. We'll share that work when its ready, then we'll start design & concept testing early next year.

**Reply by tlambert03** [reply]
> Its important to get right, though, otherwise its either not useful or (worst case) actively harmful.

This really resonates with me... and @haesleinhuepf I think this would be my primary concern when you state

> "This train is gaining speed and we can't stop it. We are fully aware that things may break and are happy to fix them. (Did you know that Fiji breaks all the time?😉)" 

In other words, "more" and "faster" isn't _always_ better.  And I think there is a very real possibility of damaging a young community.  Even if you are willing, you may not be fully _capable_  of fixing all of the broken user expectations or community impressions that get created.  I _know_ that Fiji breaks all of the time and I routinely see the frustration that it generates in users when they feel helpless to do anything about it.  I don't expect that we can create an ecosystem that will never break... but with forethought and buy in from key developers such as yourself, we can try to improve upon the situation for the end-user (so that it's clearer what broke, and how to report it).  I really hope that you'll work with us to that aim.

**Reply by haesleinhuepf** [reply]
> In other words, "more" and "faster" isn't always better. 

I know what you mean. I just don't know how to slow down that process. Just to clarify, what I meant with acceleration. We have people from about 7 research groups here on campus who develop napari plugins at the moment. In my group it's four people including myself. In my institute, multiple research groups are founded right now and the first new postdocs arriving ask me "Shall we go for ImageJ or try napari? Shall we run our new custom microscopes with Java or Python?" What would you answer? That's what I meant with the "train". Furthermore, multiple groups ask me for "plot profile" and "cropping" tools at the same time and that's why I'm pushing out publicly available plugins such as [plot-profile](https://github.com/haesleinhuepf/napari-plot-profile) and [napari-crop](https://github.com/haesleinhuepf/napari-crop). You know, it's my responsibility to serve the needs of local scientists, especially those who can't code such plugins themselves.

**Reply by tlambert03** [reply]
> I know what you mean. I just don't know how to slow down that process.

totally get it :)  and... 

A) I love that the answer to  "Shall we run our new custom microscopes with Java or Python?" is increasingly "let's try python!".  makes my heart go flutter.
B) I love that napari is a central part of this... helps us work out the kinks much faster.
C) I love that you make these plugins (plot profile, napari crop, etc), for your local users.
D) I love that you publish them and make them public, you're a real boon to napari!

honestly,  that really is my dominant emotion in all this.  You're doing amazing stuff and I don't want to sound like I want to stop the train or anything ... just didn't know where you got the idea that non-programmers are not an important part of napari anymore, or that we're not particularly interested in the life sciences community anymore.  And I hope that we don't conflate "not stopping the train" with going off and creating a parallel plugin ecosystem cause napari's not moving fast enough.   You're a very visible member out there, and "with great power comes great responsibility" 😂 

thanks again!

**Reply by haesleinhuepf** [reply]
>  just didn't know where you got the idea that non-programmers are not an important part of napari anymore

It was the tone in a couple of zoom meetings 2-4 weeks ago. I explicitly asked for it and got replies like "life sciences is just one thing, napari aims at a wider community" and "focus on the python community"... I misunderstood that obviously. Btw. I also know from former projects that these misunderstandings come from lack of exchange on a personal level. Remote-collaborations suffer from that. You need to meet physically from time to time and drink beer together in order to really understand each others motivations and goals. NEUBIAS was extremely successful in this context. I hope, one day that we all meet physically and clear up these kinds of misunderstandings. We should totally hava a plugin-hackathon, in Dresden for example :-)

**Reply by tlambert03** [reply]
> "life sciences is just one thing, napari aims at a wider community"

i do agree with this statement very much.  it _is_ just one thing, and i wouldn't want us to make decisions that preclude the possibility of napari being useful to, for example, the astronomy or geo communities... but it's also the specific community that is very near and dear to most of the core napari dev's hearts! :)  we love it too.

**Reply by sofroniewn** [reply]
> You need to meet physically from time to time and drink beer together in order to really understand each others motivations and goals.

Count me in!

**Comment by sofroniewn** [comment]
> I recently had some discussions with @sofroniewn, @jni and others who remineded me of the first sentence on the website: "napari is a fast, interactive, multi-dimensional image viewer for Python." Thus, I have the impression non-programmers are not an important target audience of napari (anymore?).

Hi @haesleinhuepf  - my bad if I gave that impression - it certainly wasn't my intent. There has been no change in the mission and values of napari over the past year, and if you look at those statements they clearly indicate that we want to

> providing GUI access to all critical functionality so napari can be used by people with no coding experience.

see https://napari.org/community/mission_and_values.html#our-mission.

As @tlambert03 said above though, this takes time to do, but we are trying to do it in a long term stable and sustainable way.

> I would hope that the early adopters would want to work closely with napari to identify places in which napari could do better for non coders. 

I completely agree with @tlambert03 here and now could be a good time to think about what are more of those essential features that we need to add to the core of napari, and work on contributing those to the main repo. @jni also is working on curating a list here too.

> I now realize it's possible to make meta plugins, that are really just collections of plugins, such as: devbio-napari

As to this @psobolewskiPhD. While it is possible, and would be of some immediate value, I'm not sure this approach is the best long term solution to sharing "starter packs". There have been some discussions with @neuromusic that one day the hub could allow people to share curated lists of plugins without having to create a new plugin, or share workflows that depend on plugins but aren't themselves plugins. We might go in the direction of meta plugins, but there could also be better alternatives that are simpler lists etc.

**Reply by haesleinhuepf** [reply]
> There has been no change in the mission and values of napari over the past year

Oh great! Because I really thought the team shifted from a life-science mission to a general-python mission. Apologies for my obviously wrong interpretation. Anyway, my (teams) mission focuses on what non-coders need, as we have many of them on campus, and thus, we are going in that direction even though napari is still experimental. This train is gaining speed and we can't stop it. We are fully aware that things may break and are happy to fix them. (Did you know that Fiji breaks all the time?😉) I also hope we can contribute more to the core at some point but there are some rather political barriers to be removed first. For now, it appears to us as strategic advantage to use napari and through providing plugins make it useful for our purposes as we mostly work with 3D (+ time) data, which is hard on other platforms such as ImageJ. Thanks again everyone for the fantastic platform you're crafting and maintaining 🙌

**Comment by LCObus** [comment]
Dresden plugin-hackathon Summer 2022?

On Mon, Oct 25, 2021 at 8:43 PM Nicholas Sofroniew ***@***.***>
wrote:

> You need to meet physically from time to time and drink beer together in
> order to really understand each others motivations and goals.
>
> Count me in!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/chanzuckerberg/napari-hub/discussions/299#discussioncomment-1536016>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APFJIOZ7HUUTSYVLZXVNT2LUIYPUTANCNFSM5GRK2I5Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
*Lucy Coulson Obus *(she/her)
Staff User Experience Researcher, Imaging Technology
napari.org | napari-hub.org | chanzuckerberg.com/science

**Reply by jni** [reply]
I'm in!

**Comment by psobolewskiPhD** [comment]
> Under Authors: the `Robert Haase` would be a clickable link, like all the others to the right.

The discussion derailed a bit to broader issues, but I'd like to bump this. Having the authors be clickable on a plugin page to bring up their other plugins (even if just via search) is still something I'd really like to see.

**Reply by neuromusic** [reply]
Thanks Peter! Yes, this is still on our radar... now that we've launched the Plugin Preview Page and Categories, we'll start on some of these new features soon.

**Reply by psobolewskiPhD** [reply]
What/where is the Plugin Preview page? sounds delightful.
Edit:
Also categories are great, but how to I filter? eg. workflow step brings up the submenu, but clicking on then categories doesn't do anything, even though it looks like they should be interactive (Safari macOS 12). I don't see any `apply` or `filter` button 😢

**Reply by neuromusic** [reply]
> I don't see any apply or filter button

It should filter as soon as you click out of the submenu... does that work?

**Reply by psobolewskiPhD** [reply]
click out? Not sure I get what you mean.

**Reply by psobolewskiPhD** [reply]
Oh! I see now! Click to select then **click somewhere else**.
Ok, that works, but man I can't say that behaves like any other such widget I've used before.
Especially since *clicking* on the sub-categories visually does something (they blink) but then doesn't invoke anything.
I'd prefer either the actual click on sub-category doing the action of filtering or the highlight being permanent and then adding a filter/apply button.

**Reply by neuromusic** [reply]
Ah, yes, I can see how that might not be intuitive. I've shared your feedback with the team and we'll look into it :D

**Reply by neuromusic** [reply]
Done. Check it out @psobolewskiPhD: https://www.napari-hub.org/?sort=recentlyUpdated&authors=Robert+Haase&page=1

**Comment by psobolewskiPhD** [comment]
@neuromusic I love that the author names are clickable! Thanks! I really think this is a big improvement. 
But I have to report a bug: multi-dev plugins. 
See for example:
https://www.napari-hub.org/plugins/napari-clusters-plotter
Looks like all of the author names are clumped as one? Brings up just this one plugin. I can't get `Laura Zigutyte` separately to see what other plugins she's authored...

Next, I would love if the author clicks went to a new landing page, not just filtered search. More like GitHub clicking on a dev? So there would be obviously all the plugins, but maybe also some copy about the dev, a picture, etc?
And then, maybe plugin authors could also have something like Github stars where I could check out other plugins they use/are interested in. Ideally leading to the curated meta-plugins mentioned above.

Keep up the good work!

**Reply by LCObus** [reply]
Great call out on the multi-dev plugin issue!
And are you reading our minds? Author profiles are on our horizon, and we hope our work in the new collections feature space will serve a version of the github stars need you're interested in, but ill be excited to think a little more about the stars goal.

**Reply by psobolewskiPhD** [reply]
I'm not sure how it works now, because I don't have a plugin, but maybe devs of a plugin on the hub could log in, ideally using GitHub login, like image.sc, etc. and "claim" their author/dev page? Then they could provide a photo, some copy, etc. Or maybe it could be copied over from GitHub.

**Reply by neuromusic** [reply]
> Looks like all of the author names are clumped as one? 

Indeed. The issue here is that PyPI's metadata only supports one author, which plugin devs get around by listing multiple in the field separated by commas or ampersands or "and". We explored parsing these and could technically do so, but not without clobbering authors like ["Computational Microscopy Platform, CZ Biohub"](https://www.napari-hub.org/plugins/recOrder-napari).

HOWEVER! We do have a solution...

If authors use the "Author" field in `.napari-hub/config.yml` to list multiple authors, then filtering by author will work as expected.

In the coming weeks, we're going to do some bulks PRs and will try to help plugin devs make the necessary changes to take advantage of this feature.
